src/MainWindow.vala

```diff
@@ -214,6 +214,16 @@ public class Mail.MainWindow : Hdy.ApplicationWindow {
             session_started ();
         });
 
+        delete_event.connect (() => {
+            unowned var application = (Application)GLib.Application.get_default ();
+            application.request_background.begin ((obj, res) => {
+                application.request_background.end (res);
+                destroy ();
+            });
```

Comment:
MainWindow already has a application reference, no need to call Application.get_default() and you don't need to call end() if you won't catch any error or work with the return value in async methods.
```suggestion
            application.request_background.begin (() => destroy ());
```